app/src/main/java/org/mozilla/tv/firefox/fxa/FxaLoginUseCase.kt

```diff
@@ -0,0 +1,101 @@
+    private val sessionRepo: SessionRepo,
+    private val screenController: ScreenController,
+    private val sentryIntegration: SentryIntegration = SentryIntegration
+) {
+
+    /**
+     * Opens the browser screen and loads the FxA login URL to begin the login flow.
+     */
+    fun beginLogin(fragmentManager: FragmentManager) {
+        // TODO: should we throw an error if we're already authenticated when this is called?
+        GlobalScope.launch(Dispatchers.Main) { // main thread: we modify UI state.
+            // a-c#3713: this await will never resume if the user is already logged in.
+            val loginUri = fxaRepo.beginLoginInternalAsync().await()
+
+            // This may be null when the FxA library fails to do things internally, mostly likely due to network issues.
+            if (loginUri == null) {
+                sentryIntegration.captureAndLogError(
+                    LOGTAG, IllegalStateException("beginAuthenticationAsync returned null loginUri"))
+                return@launch
+            }
+
+            // TODO: if user is already signing in, we should consider not reloading the page.
+            screenController.showBrowserScreenForUrl(fragmentManager, loginUri)
+        }
+    }
+
+    init {
+        attachFxaLoginSuccessObserver() // @Suppress cannot be used on an initializer so we call a helper method.
+    }
+
+    @Suppress("CheckResult") // no need to dispose: sessionRepo is active for the duration of the app.
+    private fun attachFxaLoginSuccessObserver() {
+        fun isLoginSuccessUri(uri: String): Boolean = uri.startsWith(FxaRepo.REDIRECT_URI)
+
+        fun extractLoginSuccessKeys(uriStr: String): LoginSuccessKeys? {
+            val uri = Uri.parse(uriStr)
+            val code = uri.getQueryParameter("code")
+            val state = uri.getQueryParameter("state")
+            return if (code != null && state != null) LoginSuccessKeys(code = code, state = state) else null
+        }
+
+        sessionRepo.state
+            .map { it.currentUrl }
+            .distinctUntilChanged()
+            .subscribe {
+                if (!isLoginSuccessUri(it)) {
+                    return@subscribe
+                }
+
+                val loginSuccessKeys = extractLoginSuccessKeys(it)
+                if (loginSuccessKeys == null) {
+                    // Since we received a login success URL, this is never expected. However, since this action is
+                    // controlled by a server, we don't want to crash the app so we log to Sentry instead.
+                    sentryIntegration.captureAndLogError(
+                        LOGTAG, IllegalStateException("Received success URI but success keys cannot be found"))
+                    return@subscribe
+                }
+
+                // TODO: do we want to pop these fxa pages from the user's browsing history?
+                fxaRepo.accountManager.finishAuthenticationAsync(loginSuccessKeys)
+            }
```

Comment:
The conditional statements in this block amount to a `filter` and `filterMap`.  `filterMap` is an FP concept, but doesn't have its own operator, we accomplish it in Rx using a `flatMap`.  We can rewrite this as:

```
fun Observable<String>.filterMapSuccessKeys(): Observable<LoginSuccessKeys> =
    this.flatMap { url ->
        val loginSuccessKeys = extractLoginSuccessKeys(url)
        
        if (loginSuccessKeys != null) {
            Observable.just(loginSuccessKeys)
        } else {
            // Since we received a login success URL, this is never expected. However, since this action is
            // controlled by a server, we don't want to crash the app so we log to Sentry instead.
            sentryIntegration.captureAndLogError(
                LOGTAG, IllegalStateException("Received success URI but success keys cannot be found"))
            Observable.empty()
        }
    }

sessionRepo.state
    .map { it.currentUrl }
    .distinctUntilChanged()
    .filter { isLoginSuccessUri(it) }
    .filterMapSuccessKeys()
    .subscribe { loginSuccessKeys ->
        // TODO: do we want to pop these fxa pages from the user's browsing history?
        fxaRepo.accountManager.finishAuthenticationAsync(loginSuccessKeys)
    }
```

Reply:
This is overall an improvement in readability: neat. :) I don't love how unintuitive flatMap as filterMap is, especially for new devs, but creating a new data type to handle the empty case sounds bad too.

I never would have guessed Rx was missing an operator. :P

Reply:
Haha, yeah.  I consider sometimes writing a small `filterMap` operator that handles this part for us.  I've been waffling back and forth on that one for a while....